haystack/agents/base.py

```diff
@@ -0,0 +1,298 @@
+        tools: Optional[List[Tool]] = None,
+        max_iterations: int = 5,
+        tool_pattern: str = r'Tool:\s*(\w+)\s*Tool Input:\s*("?)([^"\n]+)\2\s*',
+        final_answer_pattern: str = r"Final Answer:\s*(\w+)\s*",
+    ):
+        """
+         Creates an Agent instance.
+
+        :param prompt_node: The PromptNode that the Agent uses to decide which tool to use and what input to provide to it in each iteration.
+        :param prompt_template: The name of a PromptTemplate supported by the PromptNode. It is used for generating thoughts and running Tools to answer queries step-by-step.
+        :param tools: A List of Tools that the Agent can choose to run. If no Tools are provided, they need to be added with `add_tool()` before you can use the Agent.
+        :param max_iterations: The number of times the Agent can run a tool plus then once infer it knows the final answer.
+            Set at least to 2 so that the Agent can run one Tool and then infer it knows the final answer. Default 5.
+        :param tool_pattern: A regular expression to extract the name of the Tool and the corresponding input from the text generated by the Agent.
+        :param final_answer_pattern: A regular expression to extract final answer from the text generated by the Agent.
+        """
+        self.prompt_node = prompt_node
+        self.prompt_template = prompt_template
+        self.tools = {tool.name: tool for tool in tools} if tools else {}
+        self.max_iterations = max_iterations
+        self.tool_pattern = tool_pattern
+        self.final_answer_pattern = final_answer_pattern
+        send_custom_event(event=f"{type(self).__name__} initialized")
+
+    def add_tool(self, tool: Tool):
+        """
+        Add the provided tool to the Agent and update the template for the Agent's PromptNode.
+
+        :param tool: The Tool to add to the Agent. Any previously added tool with the same name will be overwritten.
+        """
+        self.tools[tool.name] = tool
+
+    def has_tool(self, tool_name: str):
+        """
+        Check whether the Agent has a Tool registered under the provided tool name.
+
+        :param tool_name: The name of the Tool for which to check whether the Agent has it.
+        """
+        return tool_name in self.tools
+
+    def run(
+        self, query: str, max_iterations: Optional[int] = None, params: Optional[dict] = None
+    ) -> Dict[str, Union[str, List[Answer]]]:
+        """
+        Runs the Agent given a query and optional parameters to pass on to the tools used. The result is in the
+        same format as a pipeline's result: a dictionary with a key `answers` containing a List of Answers
+
+        :param query: The search query.
+        :param max_iterations: The number of times the Agent can run a tool plus then once infer it knows the final answer.
+            Set at least to 2 so that the Agent can run one tool and then infer it knows the final answer. Default 5.
```

Comment:
Isn't this a bit misleading? Technically the default is whatever has been set in the `Agent` constructor, if we set it to 10 in there but leave it untouched in `run` it will default to 10, right? 🤔

Reply:
The idea is that we might want to use different values of max_iterations depending on the query but all with the same Agent. So a user can set a default when initializing the Agent and then later overwrite it.

Reply:
Ah, you mean that `Set at least to 2` is confusing because it's already set in the Agent's init. 👍 I agree.

Reply:
I updated the docstring to make clear that the parameter is optional and removed the statement about the default value.